Sources/NIOSSL/SSLContext.swift

```diff
@@ -779,6 +791,90 @@ extension NIOSSLContext {
     }
 }
 
+// For utilizing SSL Context callbacks from a SSLContext
+extension NIOSSLContext {
+    static private func setSSLContextCallback(context: OpaquePointer) throws {
+        CNIOBoringSSL_SSL_CTX_set_cert_cb(context, { (ssl, arg) in
+            guard let ssl = ssl else {
+                return SSL_TLSEXT_ERR_NOACK
+            }
+
+            let parentSwiftContext = NIOSSLContextHelpers.getContextFromRawContextAppData(ssl: ssl)
+
+            // Check if we have a pending context previously set by this callback
+            if let futureSSLContext = parentSwiftContext.futureSSLContext {
+                // If we do and we are executing this callback then the future has already resolved
+                // We can safely call wait() which will resolve immediately
+                let userChosenContext: NIOSSLContext
+                do {
+                    userChosenContext = try futureSSLContext.wait()
+                } catch {
+                    return SSL_TLSEXT_ERR_NOACK
+                }
+
+                // The user may or may not retain a reference to the NIOSSLContext
+                // that they return here, if they provide one. The NIOSSLContext finalizer
+                // explicitly decrements the reference count on the underlying SSL_CTX.
+                //
+                // However, SSL_set_SSL_CTX explicitly increases the reference count and
+                // will decrement the count when the SSL* object is destroyed, thus finally
+                // freeing the raw SSL_CTX object, even if the swift reference is orphaned.
+                //
+                // As such, no memory should be leaked or freed too early in the event that
+                // the user fails to retain their created and return NIOSSLContext. In the
+                // event that they do retain it, the reference count will remain at at-least
+                // 1 and thus they can re-use the same context for multiple connections.
+                let nativeContextValue = withExtendedLifetime(userChosenContext) {
+                    return CNIOBoringSSL_SSL_set_SSL_CTX(ssl, userChosenContext.sslContext)
+                }
+
+                // CNIOBoringSSL_SSL_set_SSL_CTX should return a pointer to the very same context it was
+                // told to assign on success. On error, this will not happen, so we check for equality
+                // here and handle any such errors by signaling the connection to fail.
+                if nativeContextValue != userChosenContext.sslContext {
+                    return SSL_TLSEXT_ERR_NOACK
+                }
+
+                return SSL_TLSEXT_ERR_OK
+            }
+
+            // Make sure we get the server name extracted.
+            guard let cServerName = CNIOBoringSSL_SSL_get_servername(ssl, CNIOBoringSSL.TLSEXT_NAMETYPE_host_name) else {
```

Comment:
Nit: we shouldn't need the explicit module name:

```suggestion
            guard let cServerName = CNIOBoringSSL_SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name) else {
```